File: Java/1282_Group_the_People_Given_the_Group_Size_They_Belong_To.java
class Solution {
    public List<List<Integer>> groupThePeople(int[] groupSizes) {
        List<List<Integer>> result = new ArrayList<>();
        List<List<Integer>> temp = new ArrayList<>();
        for(int i = 0; i < groupSizes.length; i++){
            while(temp.size() < groupSizes[i]){
                List<Integer> item = new ArrayList<>();
                temp.add(item);
            }
            
            List<Integer> item = temp.get(groupSizes[i]-1);
            
            item.add(i);
           
            if(item.size() == groupSizes[i]){
                result.add(new ArrayList(item));
                // result.add(item);
                item.clear();
            }
        }
        return result;
    }
}

File: Cpp/1282_Group_the_People_Given_the_Group_Size_They_Belong_To.cpp
class Solution {
public:
    vector<vector<int>> groupThePeople(vector<int>& groupSizes) {
        vector<vector<int>> result;
        
        unordered_map<int, vector<int>> m;
        //map<int, vector<int>> m;
        for(int i = 0; i < groupSizes.size(); i++){
            /*
            if(m.count(groupSizes[i]) == 0){   
            //if(m.find(groupSizes[i]) == m.end()){
                m.insert(pair<int, vector<int>>(groupSizes[i], vector<int>() ));
            }
            */
            m[groupSizes[i]].push_back(i);
            
            if(m[groupSizes[i]].size() == groupSizes[i]){
                result.push_back(m[groupSizes[i]]);
                m[groupSizes[i]].clear();
                //m.erase(groupSizes[i]);
            }
        }
        
        return result;
    }
};
/*

for(map<int, vector<int>>::iterator iter = m.begin(); iter != m.end(); iter++)
{
    cout<<iter->first<<endl;
    cout<<iter->second.size()<<endl;
}	

*/
